lib/health/health.go

```diff
@@ -122,32 +126,53 @@ func (aggregator *HealthAggregator) Summary() *HealthReport {
 	aggregator.mutex.Lock()
 	defer aggregator.mutex.Unlock()
 
+	logFuncs := []func(){}
+
 	// In the absence of any reporters, default to indicating that we are both live and ready.
 	summary := &HealthReport{Live: true, Ready: true}
 
 	// Now for each reporter...
 	for name, reporter := range aggregator.reporters {
-		log.WithFields(log.Fields{
-			"name":           name,
-			"reporter-state": reporter,
-		}).Debug("Detailed health state")
-
 		// Reset Live to false if that reporter is registered to report liveness and hasn't
 		// recently said that it is live.
-		if summary.Live && reporter.reports.Live && (!reporter.latest.Live ||
-			(reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)) {
+		notReportedLive := !reporter.latest.Live || (reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)
+		if summary.Live && reporter.reports.Live && notReportedLive {
 			summary.Live = false
 		}
 
 		// Reset Ready to false if that reporter is registered to report readiness and
 		// hasn't recently said that it is ready.
-		if summary.Ready && reporter.reports.Ready && (!reporter.latest.Ready ||
-			(reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)) {
+		notReportedReady := !reporter.latest.Ready || (reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)
+		if summary.Ready && reporter.reports.Ready && notReportedReady {
 			summary.Ready = false
 		}
+
+		logEntry := log.WithFields(log.Fields{
+			"name":           name,
+			"reporter-state": reporter,
+		})
+
+		if reporter.reports.Live && notReportedLive || reporter.reports.Ready && notReportedReady {
+			logFuncs = append(logFuncs, func() {
+				logEntry.Warn("Unhealthy reporter")
+			})
+		} else {
+			logFuncs = append(logFuncs, func() {
+				logEntry.Debug("Healthy reporter")
+			})
+		}
 	}
 
-	log.WithField("summary", summary).Info("Overall health")
+	// Summary status has changed so update previous status and log.
+	if summary.Live != aggregator.lastReport.Live && summary.Ready != aggregator.lastReport.Ready {
+		aggregator.lastReport = summary
+		log.WithField("lastSummary", summary).Info("Overall health")
```

Comment:
Now you've put this behind a check, how about changing the log to say "Overall health status changed"